doc/developer/design/20230607_shift_status_updates.md

```diff
@@ -0,0 +1,150 @@
+
+## Detailed description
+
+<!--
+Describe the approach in detail. If there is no clear frontrunner, feel free to
+list all approaches in alternatives. If applicable, be sure to call out any new
+testing/validation that will be required.
+
+For some features it can be helpful to sketch an implementation. If you're
+working on things that are crossing team boundaries it will be helpful to spell
+out any new interfaces/traits/interactions.
+
+For most new features, you should think about testing, rollout/lifecycle, and
+observability. These things can warrant their own sections.
+-->
+
+We will introduce a couple new pieces of state to `mz_storage::storage_state::StorageState` in order to keep track of the statuses of sink and source ingestion dataflows:
+
+```rust
+pub source_statuses: Rc<RefCell<BTreeMap<GlobalId, Option<HealthStatus>>>>,
+pub sink_statuses: Rc<RefCell<BTreeMap<GlobalId, Option<HealthStatus>>>>,
+```
+
+For source ingestion dataflows, the health operator will evolve to update the status corresponding to the source's `GlobalId` in `source_statuses`. For sinks, `Healthchecker::update_status` will evolve to do the same for `sink_statuses`.
+
+The storage worker will be updated to periodically send `StorageResponse::StatusUpdates` to the storage controller, which will be defined as:
+
+```rust
+pub enum StorageResponse {
+    ...,
+    StatusUpdates(Vec<ObjectStatusUpdate>),
+}
+
+pub enum ObjectStatusUpdate {
+    Sink(SinkStatusUpdate),
+    Source(SourceStatusUpdate),
+}
+
+// Same for `SinkStatusUpdate`
+pub struct SourceStatusUpdate {
+    id: GlobalId,
+    status: String,
+    error: String,
+    hint: Option<String>,
+}
+```
+
+To avoid sending this information on every tick of the storage worker's main processing loop, we will only send updates per some defined frequency (e.g. every 5 seconds). This is similar to how statistics reporting is currently implemented. Upon receiving a `StorageResponse::StatusUpdates` message, the storage controller will append the status updates to the appropriate managed collection. Writes to the backing persist shard will take place in a dedicated task, namely the one spawned by `mz_storage_client::controller::collection_mgmt::CollectionManager`.
+
+When dropping a replica, the storage controller will need to retrieve all source/sink dataflows associated with the given storage instance. To achieve this, we can keep a mapping of storage instance id to active ingestions/exports on that instance as part of the storage controller's state. Upon creating collections/exports we can update the map and upon dropping a storage instance, we remove the corresponding entry in the map. The storage controller will only write statuses upon receiving a `StorageResponse::StatusUpdates` response iff the specified `GlobalId` exists in some active mapping. This allows us to ignore any late status updates that are sent back from storage workers.
```

Comment:
I _think_ this is exactly the state that the `RehydratingStorageClient` stores. And when the storage controller drops a `RehydratingStorageClient`, it can be sure that it won't receive any more responses from that storage instance (because the client is the means by which it was receiving those responses in the first place).

Reply:
> when the storage controller drops a RehydratingStorageClient, it can be sure that it won't receive any more responses from that storage instance (because the client is the means by which it was receiving those responses in the first place).

Right, good callout!

> I think this is exactly the state that the RehydratingStorageClient stores

I believe this state is currently obscured given that it lives in `RehydrationTask`. We could lift it up into the `RehydrationStorageClient` and still maintain that `RehydrationTask` is the source of truth for writing down all observed sources/sinks for a particular storage instance. That should give us a mechanism to access the underlying dataflow identifiers so we can set their statuses to `paused` upon dropping the backing instance.